flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializerGenerator.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.typeutils.runtime;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+
+import org.apache.flink.util.InstantiationUtil;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.api.java.typeutils.PojoTypeInfo.accessStringForField;
+import static org.apache.flink.api.java.typeutils.PojoTypeInfo.modifyStringForField;
+
+public final class PojoSerializerGenerator<T> {
+	private static final String packageName = "org.apache.flink.api.java.typeutils.runtime.generated";
+
+	private final Class<T> clazz;
+	private final Field[] refFields;
+	private final TypeSerializer<?>[] fieldSerializers;
+	private final ExecutionConfig config;
+	private String code;
+
+	public PojoSerializerGenerator(
+		Class<T> clazz,
+		TypeSerializer<?>[] fields,
+		Field[] reflectiveFields,
+		ExecutionConfig config) {
+		this.clazz = checkNotNull(clazz);
+		this.refFields = checkNotNull(reflectiveFields);
+		this.fieldSerializers = checkNotNull(fields);
+		this.config = checkNotNull(config);
+		for (int i = 0; i < this.refFields.length; i++) {
+			this.refFields[i].setAccessible(true);
+		}
+	}
+
+	public TypeSerializer<T> createSerializer()  {
+		final String className = clazz.getCanonicalName().replace('.', '_') + "_GeneratedSerializer";
+		final String fullClassName = packageName + "." + className;
+		code = InstantiationUtil.getCodeForCachedClass(fullClassName);
+		if (code == null) {
+			generateCode(className);
+		}
+		return new GenTypeSerializerProxy<>(clazz, fullClassName, code, fieldSerializers, config);
+	}
+
+	private void generateCode(String className) {
+		assert fieldSerializers.length > 0;
+		String typeName = clazz.getCanonicalName();
+		StringBuilder members = new StringBuilder();
+		for (int i = 0; i < fieldSerializers.length; ++i) {
+			members.append(String.format("final TypeSerializer f%d;\n", i));
```

Comment:
You could specialize the type here to the appropriate descendant of `TypeSerializer`. This will very likely make method calls on `f%d` faster, because of better devirtualization.